JuliaStream.jl/src/AMDGPUStream.jl

```diff
@@ -0,0 +1,178 @@
+  AMDGPU.fill!(data.a, init[1])
+  AMDGPU.fill!(data.b, init[2])
+  AMDGPU.fill!(data.c, init[3])
+end
+
+function copy!(data::ROCData{T}) where {T}
+  function kernel(a, c)
+    i = (workgroupIdx().x - 1) * workgroupDim().x + workitemIdx().x
+    @inbounds c[i] = a[i]
+    return
+  end
+  hard_wait(@roc groupsize = TBSize gridsize = gridsize(data) kernel(data.a, data.c))
+end
+
+function mul!(data::ROCData{T}) where {T}
+  function kernel(b, c, scalar)
+    i = (workgroupIdx().x - 1) * workgroupDim().x + workitemIdx().x
+    @inbounds b[i] = scalar * c[i]
+    return
+  end
+  hard_wait(
+    @roc groupsize = TBSize gridsize = gridsize(data) kernel(data.b, data.c, data.scalar)
+  )
+end
+
+function add!(data::ROCData{T}) where {T}
+  function kernel(a, b, c)
+    i = (workgroupIdx().x - 1) * workgroupDim().x + workitemIdx().x
+    @inbounds c[i] = a[i] + b[i]
+    return
+  end
+  hard_wait(
+    @roc groupsize = TBSize gridsize = gridsize(data) kernel(data.a, data.b, data.c)
+  )
+end
+
+function triad!(data::ROCData{T}) where {T}
+  function kernel(a, b, c, scalar)
+    i = (workgroupIdx().x - 1) * workgroupDim().x + workitemIdx().x
+    @inbounds a[i] = b[i] + (scalar * c[i])
+    return
+  end
+  hard_wait(
+    @roc groupsize = TBSize gridsize = gridsize(data) kernel(
+      data.a,
+      data.b,
+      data.c,
+      data.scalar,
+    )
+  )
+end
+
+function nstream!(data::ROCData{T}) where {T}
+  function kernel(a, b, c, scalar)
+    i = (workgroupIdx().x - 1) * workgroupDim().x + workitemIdx().x
+    @inbounds a[i] += b[i] + scalar * c[i]
+    return
+  end
+  hard_wait(
+    @roc groupsize = TBSize gridsize = gridsize(data) kernel(
+      data.a,
+      data.b,
+      data.c,
+      data.scalar,
+    )
+  )
+end
+
+function dot(data::ROCData{T}) where {T}
+  function kernel(a, b, size, partial)
+    tb_sum = ROCDeviceArray((TBSize,), alloc_local(:reduce, T, TBSize))
+    local_i = workitemIdx().x
+    @inbounds tb_sum[local_i] = 0.0
+
+    # do dot first
+    i = (workgroupIdx().x - 1) * workgroupDim().x + workitemIdx().x
+    while i <= size
+      @inbounds tb_sum[local_i] += a[i] * b[i]
+      i += TBSize * DotBlocks # XXX don't use (workgroupDim().x * gridDimWG().x) here
+    end
+
+    # then tree reduction
+    offset = workgroupDim().x ÷ 2
+    while offset > 0
+      sync_workgroup()
+      if (local_i - 1) < offset
+        @inbounds tb_sum[local_i] += tb_sum[local_i+offset]
+      end
+      offset ÷= 2
+    end
+
+    if (local_i == 1)
+      @inbounds partial[workgroupIdx().x] = tb_sum[local_i]
+    end
+
+    return
+  end
+  partial_sum = ROCArray{T}(undef, DotBlocks)
+  hard_wait(
+    @roc groupsize = TBSize gridsize = TBSize * DotBlocks kernel(
+      data.a,
+      data.b,
+      data.size,
+      partial_sum,
+    )
+  )
+  return sum(partial_sum)
+end
+
+function read_data(data::ROCData{T})::VectorData{T} where {T}
+  return VectorData{T}(data.a, data.b, data.c, data.scalar, data.size)
+end
+
+main()
```

Comment:
Do we need `main()` here?

Reply:
It's needed otherwise nothing runs when you load the script, `main()` isn't a special function the the runtime calls for you, see https://stackoverflow.com/a/35387566/896997

Reply:
PS: This implementation isn't a module and can't be used as one so we're not using the `__init__`